crates/worker/src/partition/state_machine/command_interpreter/mod.rs

```diff
@@ -842,6 +890,17 @@ where
                             &invocation_target,
                             idempotency_key,
                         ));
+
+                        // For workflow, we can now cleanup the service lock.
+                        if invocation_target.invocation_target_ty()
+                            == InvocationTargetType::Workflow(WorkflowHandlerType::Workflow)
+                        {
+                            effects.unlock_service_id(
+                                invocation_target
+                                    .as_keyed_service_id()
+                                    .expect("Workflow methods must have keyed service id"),
+                            );
```

Comment:
As a side note: Introducing special cases makes it a tad bit harder to follow what is happening in the state machine. For example, I was stumbling upon this call because I thought that we will unlock the service when the inbox gets popped and there are no more invocations. However, this is not happening because we are only doing it for exclusive virtual objects.

Reply:
there is not much i can act on for this comment i think. I hope once we get rid of the effect thing, it will be much easier to track all the state machine read/writes and all those special casings.

Reply:
I think it is not necessarily the effects but the different cases that we handle slightly differently. If we could unify the cases, then things would become simpler, I believe.

Reply:
> If we could unify the cases, then things would become simpler, I believe.

As much as I agree with you, right now I unfortunately don't see this happening. The different semantics we want to support are inherently different... I think we should instead strive to make the "capabilities" of the state machine more and more modular and mixable between each other, so that the state machine reasons more about "This invocation needs feature x,  y, z" rather than "this invocation needs these behaviours because it's a workflow call". A bit in a rule engine fashion, if you get what i mean. Testability could also gain from this a lot.